### PR TITLE
[iOS] support uri file path

### PR DESCRIPTION
**Pull request by cuongchau93**
added button in example for testing the feature

### Issue:
in IOS, it currently only support assets with path
`assets-library://asset/asset.JPG?id=B84E8479-475C-4727-A4A4-B77AA9980897&ext=JPG
`

However, in some case the file path is not under asset-library, 
if we supply it with uri file path such as
`/Users/cuongchau93/Library/Developer/CoreSimulator/Devices/C1105148-CD13-4EB7-98DB-B2A631B5B46E/data/Containers/Data/Application/A298452C-613C-46B0-BC9B-11FDD1475C1B/tmp/DEFA99E7-CA1F-4A77-A16B-4321CDCE3DA8.jpg
`
it does not work (in my app, it crashed. In the example app, it returns black as dominant color)

### CODE REVIEW

File: ios/RNPalette.m [hunk 1]
```diff
@@ -6,20 +6,14 @@ @implementation RNPalette
...
-  float dimension = [RCTConvert float:options[@"dimension"]]; // 4
-  float flexibility = [RCTConvert float:options[@"flexibility"]]; // 5;
-  float range = [RCTConvert float:options[@"range"]]; // 40;
-
-
-  [library assetForURL:aURL resultBlock:^(ALAsset *asset) {
-    UIImage  *image = [UIImage imageWithCGImage:[[asset defaultRepresentation] fullScreenImage] scale:0.5 orientation:UIImageOrientationUp];
-
-
+- (void) extractColor: (UIImage*) image
```
Comment (cuongchau93) [inline ios/RNPalette.m:9]: this function is extracted from previous version, no change to this function

File: index.js [hunk 2]
```diff
@@ -24,74 +24,86 @@ function computeTextColor(uicolors) {
...
-        swatches.push({color: rgba,
-                       population: res[UIColor],
-                       bodyTextColor: textColor,
-                       titleTextColor: textColor,
-                       swatchInfo: UIColor,
+        const r = uicolors[0];
+        const g = uicolors[1];
+        const b = uicolors[2];
+        const a = uicolors[3];
+        const hex = `#${toHex(r * 255)}${toHex(g * 255)}${toHex(b * 255)}${toHex(0.6)}`;
```
Comment (chetstone) [inline index.js:95]: Why the change from `toHex(a*255)` to `toHex(0.6)`?

**Comment by chetstone** [comment]
Thanks so much for redoing the PR to make it easier to review. 
So as I understand it, the main change is that on iOS you can now use either file path to the image or the asset-library URI. However, the image must still be in the asset-library, images cannot be accessed from arbitrary paths. This means the README should be updated to indicate that. Also the Swatch fields spec in the README should be updated to include the hex field.

File: index.js (hunk 2 above)
Comment (cuongchau93) [reply index.js:95]: Because in my case,  I only need 60% alpha
I can make it as one of available in options, (called opacity)
What do you think?

File: index.js [hunk 3]
```diff
@@ -24,74 +24,86 @@ function computeTextColor(uicolors) {
...
+    return RNPalette.getAllSwatches(image, (err, res) => {
+      if (err) {
+        return callback(err);
+      }
+      res.sort((a, b) => b.population - a.population);
+      const splitColor = res[0].color.split(',');
+      const r = Number.parseFloat(splitColor[0].split('rgba(')[1]);
+      const g = Number.parseFloat(splitColor[1]);
+      const b = Number.parseFloat(splitColor[2]);
+      res[0].hex = `#${toHex(r)}${toHex(g)}${toHex(b)}${toHex(0.6)}`;
```
Comment (cuongchau93) [inline index.js:61]: @chetstone 
side note here: in my case I only need dominant hex, that's why in this Android case, only res[0] has hex property
Should we include hex property for all swatches?

**Comment by cuongchau93** [comment]
> the main change is that on iOS you can now use either file path to the image or the asset-library URI. However, the image must still be in the asset-library, images cannot be accessed from arbitrary paths. This means the README should be updated to indicate that. Also the Swatch fields spec in the README should be updated to include the

So in my project, the images are app-only assets ( not under phone asset-library. It is under "specific app's asset-library", not sure it is called like that) 
Please check my questions, I'll start updating missing pieces after your confirmation

File: index.js (hunk 2 above)
Comment (chetstone) [reply index.js:95]: I think we should return the same thing for alpha as is returned by the native library (a*255). I don't think the extra complication of an option for opacity is necessary since it's trivial to modify the value returned by `react-native-palette`. Besides, it wouldn't support all use cases, for example, if rather than a fixed opacity (0.6) you wanted half of whatever opacity the library returned. Then you'd still have to modify the returned value.

File: index.js (hunk 3 above)
Comment (chetstone) [reply index.js:61]: Yes, if we return a hex property it should be done everywhere.

**Comment by chetstone** [comment]
> > the main change is that on iOS you can now use either file path to the image or the asset-library URI. However, the image must still be in the asset-library, images cannot be accessed from arbitrary paths. This means the README should be updated to indicate that. Also the Swatch fields spec in the README should be updated to include the
> 
> So in my project, the images are app-only assets ( not under phone asset-library. It is under "specific app's asset-library", not sure it is called like that)

So are your images in the included [via Xcode asset catalogs or in the Android drawable folder](https://reactnative.dev/docs/images#images-from-hybrid-apps-resources)? Or are they [in the javascript tree](https://reactnative.dev/docs/images#static-image-resources)? Or somewhere else? Which ones work and which don't work? It would be nice to be able to document a little more precisely what happens.

Thanks a lot for following up on this.

Cheers

**Comment by cuongchau93** [comment]
@chetstone The image file is fetched from remote server and saved to device storage
The file path in description is returned from [react-native-fs](https://www.npmjs.com/package/react-native-fs)


`RNFS.LibraryDirectoryPath `

If you looked in the ios part, only images saved under assets-library:// are supported
But with this PR, images with  absolute path to the NSLibraryDirectory also supported

**Comment by cuongchau93** [comment]
@chetstone updated readme and pending comments, please check, 
one question though, locally, i have it running with newer react native, should i push the updated package.json devDepenedencies ?

**Comment by MrOggy85** [comment]
This looks good! I need this for my project too.

**Comment by chetstone** [comment]
Thanks for following up on this. It's not clear to me why the new Alpha option is applied only to the hex output, not the rgba output. But I think since we've been working on this for quite a while, I'll just go ahead and merge the PR and document the inconsistency in the README.

As far as react and RN devDependencies, I don't know why those are there. I could be wrong, but I think we can delete them from package.json. The only thing you need those dependencies for is if you're building the example app, in which case the dependencies are correctly mentioned in `example/package.json`. If someone wants to update those, they can submit a separate PR.
